homeassistant/scripts/auth.py

```diff
@@ -32,54 +35,124 @@ def run(args):
+    parser_enable_mfa.add_argument('password', type=str)
+    parser_enable_mfa.set_defaults(func=enable_mfa)
+
     args = parser.parse_args(args)
     loop = asyncio.get_event_loop()
     hass = HomeAssistant(loop=loop)
     hass.config.config_dir = os.path.join(os.getcwd(), args.config)
+
+    provider_config = [{'type': 'homeassistant'}]
+    module_config = [{'type': 'totp'}]
+    hass.auth = loop.run_until_complete(
+        auth.auth_manager_from_config(hass, provider_config, module_config))
+
     data = hass_auth.Data(hass)
     loop.run_until_complete(data.async_load())
-    loop.run_until_complete(args.func(data, args))
+    loop.run_until_complete(args.func(hass, data, args))
 
 
-async def list_users(data, args):
+async def list_users(hass, data, args):
     """List the users."""
     count = 0
-    for user in data.users:
-        count += 1
-        print(user['username'])
+    if args.all:
+        # pylint: disable=protected-access
+        for user in await hass.auth._store.async_get_users():
+            print("{}{}{}".format(
+                str(user.name).ljust(20),
+                str(user.id).ljust(34),
+                str(user.mfa_modules)
+            ))
+            count += 1
+            for cred in user.credentials:
+                print("  - {}".format(cred.data.get('username')))
+
+    else:
+        provider = hass.auth.auth_providers[0]
+        for user in await provider.async_credentials():
+            count += 1
+            print(user.data.get('username'))
 
     print()
     print("Total users:", count)
 
 
-async def add_user(data, args):
+async def add_user(hass, data, args):
     """Create a user."""
-    data.add_user(args.username, args.password)
+    data.add_auth(args.username, args.password)
     await data.async_save()
+
+    provider = hass.auth.auth_providers[0]
+    credential = await provider.async_get_or_create_credentials(
+        {'username': args.username})
+    await hass.auth.async_get_or_create_user(credential)
+
+    # Flush AuthStore
+    await hass.async_stop()
+
     print("User created")
 
 
-async def validate_login(data, args):
+async def validate_login(hass, data, args):
     """Validate a login."""
     try:
         data.validate_login(args.username, args.password)
-        print("Auth valid")
-    except hass_auth.InvalidAuth:
+        if args.code:
+            provider = hass.auth.auth_providers[0]
+            credential = await provider.async_get_or_create_credentials(
+                {'username': args.username})
+            user = await hass.auth.async_get_or_create_user(credential)
+
+            module = await hass.auth.async_get_auth_module('totp')
+            result = await module.async_validation_flow(
+                user.id, {'code': args.code})
+            if result is not None:
+                print("Auth valid")
+            else:
+                print("Auth invalid")
+        else:
+            print("Auth valid")
+    except auth.InvalidAuth:
```

Comment:
What was raising this? And why? Please split the mfa part away from the valid auth part.